Generate docs sidebar sections automatically

ptgott opened the pull request:
Add a `config.json` field within each `navigation` entry called `generateFrom`. This designates a relative path from `docs/pages` from which to generate entries within the sidebar.

When generating pages, a function called `generateNavPaths` looks for pages to use as second-level section introductions. These can either be called `introduction.mdx` or have the same name as a second-level subdirectory.

Adding a `generateFrom` field is consistent with the Docusaurus approach to sidebar generation, in which a configuration field indicates which directory to generate a section from. This gives us control over the title and icons we use for navigation sections, which aren't available to fetch from a directory tree alone. It also lets us use the current, hardcoded `entries` approach for some sections if we need to.

Also un-skips some accidentally skipped tests.

avatus commented on server/config-docs.ts:
```diff
@@ -327,5 +328,18 @@ export const loadConfig = (version: string) => {
   validateConfig<Config>(validator, config);
 
+  config.navigation.forEach((item, i) => {
+    if (!!item.generateFrom && item.entries.length > 0) {
+      throw "a navigation item cannot contain both generateFrom and entries";
+    }
```
thank you for the check here!

avatus commented on server/pages-helpers.ts:
```diff
@@ -61,3 +61,82 @@ export const getPageInfo = <T = MDXPageFrontmatter>(
+export const generateNavPaths = (fs, dirPath) => {
+  const firstLvl = fs.readdirSync(dirPath, "utf8");
+  let result = [];
+  let firstLvlFiles = new Set();
+  let firstLvlDirs = new Set();
```
`new Set()` 😎

ptgott commented on server/config-docs.ts:
```diff
@@ -8,11 +8,12 @@ import type { Redirect } from "next/dist/lib/load-custom-routes";
-import { resolve, join } from "path";
-import { existsSync, readFileSync } from "fs";
+import { dirname, resolve, join } from "path";
+import fs from "fs";
```
`generateNavPaths` takes `fs` as an argument so I can use the in-memory `fs` from the `memfs` package in tests!

avatus replied:
Oh I see now. That's awesome

ptgott commented:
@avatus I've reworked this a little bit to change the expectations for subdirectory pages. I've added the expectation from Docusaurus that a category page have the same title as the parent directory ([docs](https://docusaurus.io/docs/sidebar/autogenerated#category-index-convention)). Docusaurus also allows category pages to be called `index.mdx` and `README.mdx`, but Alexey advised against using `index.mdx` for non-root pages and `README.mdx` doesn't seem appropriate.
